### DISCUSSION TITLE
Remapping Copilot Key on Laptop

### DISCUSSION THREAD

**Discussion by winkelnp** [hyprwm/Hyprland#8070] [Q&A] [answered]
I recently bought a new laptop and (ofc) immediately nuked the windows on it. As all the new windows laptops now replace the right control key with a copilot key (Super Shift F23), I would like to rebind that back to right control.

A complicating factor was that F23 apparently isn't in the standard keymap anymore so `dmesg` was giving me something along the lines of "key not recognised pls give it a keycode with `setkeycodes 6e <keycode>`"
I did this, checked to see that I wasn't overwriting anything and gave it the keycode 97, which is the keycode after F12 that is empty in the standard keymap.

I then set the keybind in `hyprland.conf` with `bind = $mainMod SHIFT, code:97, exec wtype -k Control_R` thinking that would just send right control in place of that whole combo.
Instead what it's doing is sending SUPER SHIFT R_CTRL for some reason.

Does anyone have any clue what I could do?

**Comment by Aqa-Ib** [comment]
Try https://github.com/rvaiya/keyd

**Reply by winkelnp** [reply]
Thanks for the speedy reply! I can't see a way of using a keycode, they seem to use names for all the keys in the examples, do you know what  I could do there?

**Reply by Aqa-Ib** [reply]
Right, if I were in your situation I would open an issue at keyd, the developer is helpful.

**Reply by Aqa-Ib** [reply]
Look https://github.com/rvaiya/keyd/issues/825

**Reply by winkelnp** [reply]
Thanks, I've sent in an issue about the keycodes thing, I'll try that solution in the mean time :)

**Comment by Aqa-Ib** [comment]
Could this work though? `bind = $mainMod SHIFT, code:97, exec wtype -m shift -m super -k Control_R`

**Reply by winkelnp** [reply]
It seems like it should but somehow doesn't, `wev` now just shows this:
```
[14:     wl_keyboard] key: serial: 17390; time: 5787703; key: 133; state: 1 (pressed)
                      sym: Super_L      (65515), utf8: ''
[14:     wl_keyboard] modifiers: serial: 0; group: 16
                      depressed: 00000040: Mod4 
                      latched: 00000000
                      locked: 00000010: Mod2 
[14:     wl_keyboard] key: serial: 17392; time: 5787704; key: 50; state: 1 (pressed)
                      sym: Shift_L      (65505), utf8: ''
[14:     wl_keyboard] modifiers: serial: 0; group: 16
                      depressed: 00000041: Shift Mod4 
                      latched: 00000000
                      locked: 00000010: Mod2 
[14:     wl_keyboard] key: serial: 17394; time: 5787867; key: 50; state: 0 (released)
                      sym: Shift_L      (65505), utf8: ''
[14:     wl_keyboard] modifiers: serial: 0; group: 16
                      depressed: 00000040: Mod4 
                      latched: 00000000
                      locked: 00000010: Mod2 
[14:     wl_keyboard] key: serial: 17396; time: 5787872; key: 133; state: 0 (released)
                      sym: Super_L      (65515), utf8: ''
[14:     wl_keyboard] modifiers: serial: 0; group: 16
                      depressed: 00000000
                      latched: 00000000
                      locked: 00000010: Mod2 
```
whereas without the `-m` thingys there's this in the middle:
```
[14:     wl_keyboard] repeat_info: rate: 25 keys/sec; delay: 600 ms
[14:     wl_keyboard] keymap: format: 1 (xkb v1), size: 71913
[14:     wl_keyboard] repeat_info: rate: 25 keys/sec; delay: 600 ms
[14:     wl_keyboard] keymap: format: 1 (xkb v1), size: 71913
[14:     wl_keyboard] repeat_info: rate: 25 keys/sec; delay: 600 ms
[14:     wl_keyboard] keymap: format: 1 (xkb v1), size: 22573
[14:     wl_keyboard] key: serial: 18380; time: 0; key: 9; state: 1 (pressed)
                      sym: Control_R    (65508), utf8: ''
[14:     wl_keyboard] key: serial: 18381; time: 0; key: 9; state: 0 (released)
                      sym: Control_R    (65508), utf8: ''
[14:     wl_keyboard] repeat_info: rate: 25 keys/sec; delay: 600 ms
[14:     wl_keyboard] keymap: format: 1 (xkb v1), size: 71913
```

which in some ways makes sense, as the system does see super and shift going down, I don't think it will be possible to have hyprland rebind that, right?
(interestingly rebinding it works a charm in windows powertoys...)

**Reply by Aqa-Ib** [reply]
sorry, my bad, wtype is giving this error:
```
wtype -m shift -m super -k Control_R
Invalid modifier name 'super'
```
let see how it names the super modifier. edit: may be win

**Reply by winkelnp** [reply]
`win` apparently according to the manpage

**Reply by Aqa-Ib** [reply]
may be ? wtype -m shift -m win -k Control_R

**Reply by winkelnp** [reply]
That just adds this in the middle of the block:
```
[14:     wl_keyboard] modifiers: serial: 0; group: 0
                      depressed: 00000041: Shift Mod4 
                      latched: 00000000
                      locked: 00000000
[14:     wl_keyboard] modifiers: serial: 0; group: 0
                      depressed: 00000041: Shift Mod4 
                      latched: 00000000
                      locked: 00000000
[14:     wl_keyboard] modifiers: serial: 0; group: 0
                      depressed: 00000041: Shift Mod4 
                      latched: 00000000
                      locked: 00000000
```
i.e. after the delay thingys and before it sees the Control_R keypress

**Reply by winkelnp** [reply]
using `-m` to release the modifier status doesn't seem to be doing anything, while using `-p` to just release the key aborts the whole thing

I think what must be happening is that releasing the modifier in `wtype` is being overridden by the fact that the system sees that the key is still pressed and is adding the status right back. This would explain why there are so many modifier events happening in `wev`?

**Reply by Aqa-Ib** [reply]
Try: 
`bind = $mainMod SHIFT, code:97, exec wtype -M Control_R`

**Reply by winkelnp** [reply]
It errors saying that Control_R isn't a valid modifier

**Reply by winkelnp** [reply]
Is it maybe an issue that the system sees the modifier `win` as `Mod4`? or would that be irrelevant and translated by `wtype`?

edit: nope not the issue, works fine on other things.
Interestingly: if I rebind e.g. F12 to `wtype -M win -k r -m win` it doesn't show up in `wev` (like other keybindings in hypr), but it also doesn't show my run launcher

**Reply by Aqa-Ib** [reply]
> using `-m` to release the modifier status doesn't seem to be doing anything, while using `-p` to just release the key aborts the whole thing
> 
> I think what must be happening is that releasing the modifier in `wtype` is being overridden by the fact that the system sees that the key is still pressed and is adding the status right back. This would explain why there are so many modifier events happening in `wev`?

Yes I am also testing it and I think that this is the case.

**Comment by Aqa-Ib** [comment, marked as answer]
There is a person in reddit who achieved this using evsieve: https://www.reddit.com/r/linuxquestions/comments/1dnh6ge/comment/lhkoy5h/?utm_source=share&utm_medium=web3x&utm_name=web3xcss&utm_term=1&utm_content=share_button

**Reply by winkelnp** [reply]
That works a like a charm :)
not sure how I feel relying on an aur package to be running all the time my system is active though...

**Reply by Aqa-Ib** [reply]
you can download the pkgbuild and manage it yourself

**Reply by Aqa-Ib** [reply]
glad it worked though, enjoy!

**Reply by winkelnp** [reply]
Thank you so much :))

**Reply by Aqa-Ib** [reply]
you are welcome, have a nice one!

**Comment by Fadilix** [comment]
Hey guys. I just ran into the same problem with my new pc
this is what worked for me
It's in the [readme of my dotfiles](https://github.com/Fadilix/dotfiles) if you don't wanna type it all
<img width="1133" height="891" alt="image" src="https://github.com/user-attachments/assets/534765a6-1684-4e94-b1e4-d957f64ab552" />
